mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir

```diff
@@ -108,3 +108,71 @@ func.func @vector_transfer_read_i4(%arg1: index, %arg2: index) -> vector<8xi4> {
 //      CHECK32:   %[[INDEX:.+]] = affine.apply #[[MAP]]()[%[[ARG0]], %[[ARG1]]]
 //      CHECK32:   %[[VEC:.+]] = vector.transfer_read %[[ALLOC]][%[[INDEX]]], %[[PAD]] : memref<3xi32>, vector<1xi32>
 //      CHECK32:   %[[VEC_I4:.+]] = vector.bitcast %[[VEC]] : vector<1xi32> to vector<8xi4>
+
+// -----
+
+func.func @vector_store_i8(%arg0: vector<8xi8>, %arg1: index, %arg2: index) {
+    %0 = memref.alloc() : memref<4x8xi8>
+    vector.store %arg0, %0[%arg1, %arg2] :memref<4x8xi8>, vector<8xi8>
+    return
+}
+
+// Expect no conversions, i8 is supported.
+//      CHECK: func @vector_store_i8
+//      CHECK: %[[ALLOC:.+]] = memref.alloc() : memref<4x8xi8>
+//      CHECK: vector.store %[[ARG0]], %[[ALLOC:.+]][%[[ARG1]], %[[ARG2]]] : memref<4x8xi8>, vector<8xi8>
+
+//  CHECK32-DAG: affine_map<()[s0, s1] -> (s0 * 2 + s1 floordiv 4)>
+//      CHECK32: func @vector_store_i8
+//      CHECK32: %[[ALLOC:.+]] = memref.alloc() : memref<8xi32>
+//      CHECK32: %[[INDEX:.+]] = affine.apply #[[MAP]]()[%[[ARG1]], %[[ARG2]]]
+//      CHECK32: %[[VEC_I32:.+]] = vector.bitcast %[[ARG0]] : vector<8xi8> to vector<2xi32>
+//      CHECK32: vector.store %[[VEC_I32:.+]], %[[ALLOC:.+]][%[[INDEX:.+]]] : memref<8xi32>, vector<2xi32
+
+// -----
+
+func.func @vector_store_i4(%arg0: vector<8xi4>, %arg1: index, %arg2: index) {
+    %0 = memref.alloc() : memref<4x8xi4>
+    vector.store %arg0, %0[%arg1, %arg2] :memref<4x8xi4>, vector<8xi4>
+    return
+}
+
+//  CHECK-DAG: #[[MAP:.+]] = affine_map<()[s0, s1] -> (s0 * 4 + s1 floordiv 2)>
+//      CHECK: func @vector_store_i4
+//      CHECK: %[[ALLOC:.+]] = memref.alloc() : memref<16xi8>
+//      CHECK: %[[INDEX:.+]] = affine.apply #[[MAP]]()[%[[ARG1]], %[[ARG2]]]
+//      CHECK: %[[VEC_I8:.+]] = vector.bitcast %[[ARG0]] : vector<8xi4> to vector<4xi8>
+//      CHECK: vector.store %[[VEC_I8:.+]], %[[ALLOC:.+]][%[[INDEX:.+]]] : memref<16xi8>, vector<4xi8>
+
+//  CHECK32-DAG: #[[MAP:.+]] = affine_map<()[s0, s1] -> (s0 + s1 floordiv 8)>
+//      CHECK32: func @vector_store_i4
+//      CHECK32: %[[ALLOC:.+]] = memref.alloc() : memref<4xi32>
+//      CHECK32: %[[INDEX:.+]] = affine.apply #[[MAP]]()[%[[ARG1]], %[[ARG2]]]
+//      CHECK32: %[[VEC_I32:.+]] = vector.bitcast %[[ARG0]] : vector<8xi4> to vector<1xi32>
+//      CHECK32: vector.store %[[VEC_I32:.+]], %[[ALLOC:.+]][%[[INDEX:.+]]] : memref<4xi32>, vector<1xi32>
+
+// -----
+
+func.func @vector_store_i4_dynamic(%arg0: vector<8xi4>, %arg1: index, %arg2: index, %arg3: index, %arg4: index) {
```

Comment:
I think this mostly looks good. It would be good to have an example of dynamic offsets as well. So try a 1D memref of sub-byte types. You can use a `memref.subview` to introduce a dynamic offset (like this https://github.com/llvm/llvm-project/blob/ef222988b477bf91fe3b6e9cd7d881d19af2d605/mlir/test/Dialect/MemRef/emulate-narrow-type.mlir#L163 but with a dynamic offset value). So `alloc -> subview -> vector.store`.

Reply:
+1

Reply:
So, when I am trying to add the dynamic offset case, I am running into this error with the memref subview: 

int64_t mlir::detail::OffsetSizeAndStrideOpInterfaceTrait<ConcreteOp>::getStaticOffset(unsigned int) [with ConcreteOp = mlir::memref::SubViewOp; int64_t = long int]: Assertion `!(*static_cast<ConcreteOp *>(this)).isDynamicOffset(idx) && "expected static offset"' failed.

The same subview is working for the other types such as i8, i32. Is dynamic offset subview supported for subbyte memrefs? Here is the memref subview for reference: 

%arr = memref.alloc() : memref<128xi4>
%subview = memref.subview %arr[%arg1] [32] [1] : memref<128xi4> to memref<32xi4, strided<[1], offset:?>>

Reply:
What is the test case? I created one and it passes `mlir-opt` IR verification:

```mlir
func.func @foo(%arg0 : vector<4xi4>, %arg1 : index) {
  %c0 = arith.constant 0 : index
  %arr = memref.alloc() : memref<128xi4>
  %subview = memref.subview %arr[%arg1] [32] [1] : memref<128xi4> to memref<32xi4, strided<[1], offset:?>>
  vector.store %arg0, %subview[%c0] :  memref<32xi4, strided<[1], offset:?>>, vector<4xi4>
  return
}
```

Seems to be a bug in the implementation? Where is the error come from?

Reply:
Interesting, I'm not even trying to do the vector.store. It is failing at the memref.subview itself. Are you running 
`mlir-opt --test-emulate-narrow-int="memref-load-bitwidth=8" --cse --split-input-file /home/sai/llvm-project/mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir`? 
This is the test case:
```
func.func @foo(%arg1 : index) {
  %arr = memref.alloc() : memref<128xi4>
  %subview = memref.subview %arr[%arg1] [32] [1] : memref<128xi4> to memref<32xi4, strided<[1], offset:?>>
  return
}
```

Reply:
I took a look and found that the emulation pattern does not work for dynamic offsets. https://github.com/llvm/llvm-project/blame/89a2e701596835ba2714e190990da09d4e976a9a/mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp#L257-L263

According to the comment, we don't support it today. Adding such support is out of the scope of this work. We can fix it if it becomes an issue. Can you file an issue on Github with the example?

```
/// Emulating narrow ints on subview have limited support, supporting only
/// static offset and size and stride of 1. Ideally, the subview should be
/// folded away before running narrow type emulation, and this pattern would
/// never run. This pattern is mostly used for testing pruposes.
```

Reply:
Oh okay, thanks for looking into that! I will file an issue on Github with the example.

Reply:
Issue has been filed by the way